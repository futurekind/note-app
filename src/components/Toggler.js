import React, {PropTypes} from 'react';
import styled, { css } from 'styled-components';

import SectionTitle from './Sectiontitle';

const Header = styled.header`
    cursor: pointer;
`;

const Body = styled.main`
    height: 0;
    overflow: hidden;
    transform: translateY(-100px);
    opacity: 0;
    transition: transform .25s, 
                opacity .25s .1s;

    ${props => {
        if(props.open) {
            return css`
                height: auto;
                padding: 20px 0;
                transform: none;
                opacity: 1;
            `
        }
    }}
`;

const Toggler = ({
    title,
    open,
    children,
    onToggle
}) => {
    return (
        <aside>
            <Header onClick={ () => onToggle(open) }>
                <SectionTitle iconId={ open ? 'up' : 'down' }>{ title }</SectionTitle>
            </Header>

            <Body open={ open }>
                { children }
            </Body>
        </aside>
    )
}

Toggler.defaultProps = {
    open: false,
    onToggle: () => {}
}

Toggler.propTypes = {
    title: PropTypes.string.isRequired,
    open: PropTypes.bool,
    onToggle: PropTypes.func,
    children: PropTypes.any,
}

export default Toggler