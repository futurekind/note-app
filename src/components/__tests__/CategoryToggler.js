import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Toggler from '../CategoryToggler';

describe('CategoryToggler', () => {
    
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<Toggler checkboxType="green" />)
    })

    it('renders', () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
    })

    it('renders checked checkbox', () => {
        wrapper.setProps({
            checked: true
        })
        const cb = wrapper.find('Checkbox');
        
        expect(toJSON(cb)).toMatchSnapshot()
    })

    it('renders label', () => {
        wrapper.setProps({
            label: 'Some Label'
        })

        const label = wrapper.children().at(1)

        expect(toJSON(label)).toMatchSnapshot() 
    })

    it('delegates clicks on checkbox with false', () => {
        const spy = jest.fn();

        wrapper.setProps({
            onCheck: spy
        })
        
        const checkbox = wrapper.children().at(0);

        checkbox.simulate('click');
        expect(spy).lastCalledWith(false)
    })

    it('delegates clicks on checkbox with true', () => {
        const spy = jest.fn();

        wrapper.setProps({
            onCheck: spy,
            checked: true
        })

        const checkbox = wrapper.children().at(0);

        checkbox.simulate('click');
        expect(spy).lastCalledWith(true)
    })

    it('renders input field instead of label when in editMode', () => {
        wrapper.setProps({
            label: 'Some label',
            editMode: true
        })

        const label = wrapper.children().at(1)

        expect(toJSON(label)).toMatchSnapshot();
    })

    it('delegates input value on blur', () => {
        const spy = jest.fn();

        wrapper.setProps({
            label: 'Some label',
            editMode: true,
            onEditDone: spy
        })

        const input = wrapper.children().at(1).children(0)
        const event = {
            target: {
                value: 'Foo'
            }
        }

        input.simulate('blur', event);
        
        expect(spy).toBeCalledWith(event)

    })

    it('delegate click on label', () => {
        const spy = jest.fn();

        wrapper.setProps({
            label: 'Some label',
            onEdit: spy
        })

        const input = wrapper.children().at(1)

        input.simulate('click');

        expect(spy).toBeCalled()

    })
})