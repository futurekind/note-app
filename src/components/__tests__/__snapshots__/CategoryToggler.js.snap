exports[`CategoryToggler renders 1`] = `
<styled.div>
  <styled.div
    onClick={[Function]}>
    <Checkbox
      checked={false}
      onCheck={[Function]}
      type="green" />
  </styled.div>
  <styled.div
    onClick={[Function]} />
  <styled.div>
    <Icon
      id="edit"
      size={16} />
  </styled.div>
</styled.div>
`;

exports[`CategoryToggler renders checked checkbox 1`] = `
<Checkbox
  checked={true}
  onCheck={[Function]}
  type="green" />
`;

exports[`CategoryToggler renders input field instead of label when in editMode 1`] = `
<styled.div
  onClick={[Function]}>
  <styled.input
    defaultValue="Some label"
    onBlur={[Function]}
    type="text" />
</styled.div>
`;

exports[`CategoryToggler renders label 1`] = `
<styled.div
  onClick={[Function]}>
  Some Label
</styled.div>
`;
